### ISSUE TITLE
J2DBC creation operations create duplicate rows in a database

### ISSUE THREAD

**Issue by RobertBrown3** [spring-projects/spring-data-r2dbc#775]
I am seeing what I believe is a bug in the R2DBC implementation. When performing a create operation using R2DBC, the repository implementation creates duplicate rows in the database.

When creating a Spring Boot application that utilizes WebFlux, I created a relational database that I accessed using R2DBC. An example repository class, which is pretty straightforward, is provided below:

    public interface BugRepository extends ReactiveCrudRepository<DataDAO, Long>
    {

    }

An example data object designed to exercise the repositry,is provided below:

    @NoArgsConstructor
    @Data
    @Table(name = "persons")
    public class DataDAO
    {
       @Id
       private Long id;
       private String firstname;
       private String lastname;
       private String title;
   
       public DataDAO(String first,String last,String atitle)
       {
          id = null;
          firstname = first;
          lastname = last;
          title = atitle;
       }
    }

The database is a mariadb database. The SQL for creating it was:

    CREATE DATABASE IF NOT EXISTS `bugdemo`;
    USE `bugdemo`;

    CREATE TABLE IF NOT EXISTS `persons` (
      `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
      `firstname` varchar(50) DEFAULT NULL,
      `lastname` varchar(50) DEFAULT NULL,
      `title` varchar(50) DEFAULT NULL,
      PRIMARY KEY (`id`)
    ) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb4;

When writing a test that saves an instance of the DataDAO, I did the following to perform the create operation:

    @SpringBootTest
    class SavebugApplicationTests 
    {
       @Autowired
       private BugRepository theRepo;

    	@Test
    	void testAddOneRecord() 
    	{
    	   DataDAO dao = new DataDAO("Joe","User","Employee");
	   
    	   Mono<DataDAO> row = theRepo.save(dao);
	   
	       row.subscribe();
	   
	       DataDAO check = row.block();	   
	   
    	}
    }

When I call the save() method in the test, the DataDAO object gets written to the database twice. 

The save() method is always called once (I ran debugging sessions to confirm this). In an empty database, the save() operation creates two rows. Each row has the same first name, last name, and title. The difference, of course, is the ID.

I found another curious behavior: if I introduce a delay in the function (usually by setting a breakpoint on the
block() call), the second row creation fails with a duplicated record exception. 

It is almost as if the implementation of the Repository creates not one, but two threads, each of which is 
attempting the same ADD operation to the database.

I have attached a zip file containing the entire source code, configuration, and POM file (I used Maven for build
configuration, which should be easily translated into Gradle if desired). If the zip file is not sufficient, please 
let me know what would be.

NOTE: The database used with this code was MariaDB, but the duplicated rows occur when using PostgreSQL as well. I tested using Postgres, thinking that perhaps the bug was related to the MariaDB driver version. Whatever is causing this row duplication is in the R2DBC implementation.

[savebug.zip](https://github.com/spring-projects/spring-data-r2dbc/files/9276073/savebug.zip)

**Comment by mp911de** [comment]
```java
       row.subscribe();
   
       DataDAO check = row.block();
```

Initiates the save operation to happen twice. Once on `.subscribe()`, and another time on `.block()`. The publisher at `row` is a functional object that can trigger a specific action multiple times depending on how often you subscribe.

I highly recommend that you get familiar with the concepts of reactive programming. A good starting point is the Project Reactor reference documentation: https://projectreactor.io/docs/core/release/reference/

**Comment by RobertBrown3** [comment]
This was not a question of my needing to "become familiar" with reactive programming concepts. While I am admittedly a newbie,  I had read the link you provided over a week ago, and have been reading Reactor and Spring Boot Reactive examples (along with several books on the subjects) for several weeks prior. The problem is that every example I read has calls to subscribe() every time someone uses a Mono or Flux object. Problem is, if you just do a subscribe() in test applications like the one I provided, nothing happens. There is apparently no way to make a main thread hold until the action you subscribe to completes, so an application like my test application exits before the save() is completed. That is why I started adding block() calls in my test applications in the first place.

I discovered the block() method (which to my knowledge there are *no* examples using it!) and based on my experience with multithreaded programming I assumed that all it did was block until a thread was completed. Since no one actually describes "block()" as actually doing a "subscribe()" before blocking, I had no way of knowing that using it with subscribe() would cause two subscriptions to be made.

This was a question of misinterpretation of the use of a function due to incomplete documentation. I would never have even begun using Reactor- based functions if I hadn't gotten as familiar as possible with Reactive Programming "concepts" *and* the Project Reactor publisher functions. In fact, one of the books I read cover to cover was "Reactive Spring" by Josh Long, which not only gives more details about Reactive Programming "concepts" than the link you provided does, but it also gives several raw Project Reactor examples as well. So no: I am quite familiar with Reactive Programming "concepts".  To even  imply otherwise is, from my perspective, very insulting.

What I was *not* familiar with was exactly what the block() method does. As I said: I had assumed that it was a thread block, not a subscribe- and- block. There is no documentation or example that suggested otherwise.

That having been said, thank you for the information. I have other problems I need to solve using this Reactive programming library, but those I can bring to Stack Overflow or any of the other Reactive forums that I am aware of.
